### PR TITLE
Run E2E tests on Linux in build pipeline and add Backward/Forward E2E tests

### PR DESCRIPTION
**Pull request by suhsteve**
This PR runs the E2E tests on the Linux and adds Backward/Forward E2E tests

Skip 2.3.0 E2E tests running on Linux due to issue #753

### CODE REVIEW

File: azure-pipelines.yml [hunk 1]
```diff
@@ -11,8 +11,6 @@ variables:
   _SignType: real
   _TeamName: DotNetSpark
   MSBUILDSINGLELOADCONTEXT: 1
-  ArtifactPath: '$(Build.ArtifactStagingDirectory)\Microsoft.Spark.Binaries'
```
Comment (imback82) [inline azure-pipelines.yml:14]: I think this is used for internal build path?

Comment (suhsteve) [reply azure-pipelines.yml:14]: good catch.  readded

**Comment by suhsteve** [comment]
@Niharikadutta  can you look why the 2.3.0 linux test is hanging?  Looking at the other tests, ie 2.3.1/2.3.2 the next test that should be running is a test that throws that contains  `Caused by: java.io.IOException: org.apache.spark.SparkException: Attempted to use Broadcast(86) after it was destroyed (destroy at NativeMethodAccessorImpl.java:0) `

**Comment by Niharikadutta** [comment]
> @Niharikadutta can you look why the 2.3.0 linux test is hanging? Looking at the other tests, ie 2.3.1/2.3.2 the next test that should be running is a test that contains an exception messages related to broadcast `Caused by: java.io.IOException: org.apache.spark.SparkException: Attempted to use Broadcast(86) after it was destroyed (destroy at NativeMethodAccessorImpl.java:0) `

Sure I'll look into it

**Comment by Niharikadutta** [comment]
I see this warning after the test file has been triggered:
`[Warn] [JvmBridge] rmThread returned false for JVM thread 6. Either thread does not exist or has already been GC'd.
`
Is this benign?

**Comment by Niharikadutta** [comment]
I removed all Broadcast tests and saw the E2E tests for Spark 2.3.0 on Ubuntu still hanging as per #748 , so I don't think it is hanging because of Broadcast. Do we have any other ideas?

**Comment by imback82** [comment]
Can you help finding the offending test suite? Thanks!

**Comment by AFFogarty** [comment]
> I see a bunch of these warnings after the test file has been triggered:
> `[Warn] [JvmBridge] rmThread returned false for JVM thread 6. Either thread does not exist or has already been GC'd. `
> Is this benign? cc @AFFogarty

@Niharikadutta can we run the tests with Debug logging to verify that the thread GC is happening?

**Comment by suhsteve** [comment]
@imback82 , @Niharikadutta , @AFFogarty Skipping 2.3.0 E2E tests on Linux due to issue #753

**Review by Niharikadutta — COMMENTED** [review]
LGTM. Thanks @suhsteve !

**Comment by imback82** [comment]
Let me check #753 first and see if it is something that can be fixed or not.

**Comment by suhsteve** [comment]
@elvaliuliuliu  can you review?

**Comment by elvaliuliuliu** [comment]
LGTM, thanks @suhsteve!

File: azure-pipelines.yml [hunk 2]
```diff
@@ -12,13 +12,43 @@ variables:
...
+  backwardCompatibleTestOptions_Linux_2_3: ""
+  forwardCompatibleTestOptions_Linux_2_3: ""
+
+  backwardCompatibleTestOptions_Windows_2_4: ""
+  forwardCompatibleTestOptions_Windows_2_4: ""
+  backwardCompatibleTestOptions_Linux_2_4: ""
+  forwardCompatibleTestOptions_Linux_2_4: "--filter \
+  (FullyQualifiedName!=Microsoft.Spark.Extensions.Hyperspace.E2ETest.HyperspaceTests.TestExplainAPI)&\
+  (FullyQualifiedName!=Microsoft.Spark.Extensions.Hyperspace.E2ETest.HyperspaceTests.TestIndexCreateAndDelete)&\
+  (FullyQualifiedName!=Microsoft.Spark.Extensions.Hyperspace.E2ETest.HyperspaceTests.TestSignatures)"
```
Comment (imback82) [inline azure-pipelines.yml:30]: Do you know why these are filtered? I see no changes from 1.0.0?

File: azure-pipelines.yml [hunk 3]
```diff
@@ -173,18 +203,142 @@ stages:
...
-    - '3.0.1'
+    backwardCompatibleRelease: $(backwardCompatibleRelease)
+    forwardCompatibleRelease: $(forwardCompatibleRelease)
+    tests:
+    - version: '2.3.0'
+      jobOptions:
+      - pool: 'Hosted VS2017'
+        testOptions: ''
+        backwardCompatibleTestOptions: $(backwardCompatibleTestOptions_Windows_2_3)
+        forwardCompatibleTestOptions: $(forwardCompatibleTestOptions_Windows_2_3)
```
Comment (imback82) [inline azure-pipelines.yml:214]: Can you add a comment that this is disabled on ubuntu and add a link to the issue?

Comment (suhsteve) [reply azure-pipelines.yml:215]: Comment added

File: azure-pipelines.yml (hunk 2 above)
Comment (suhsteve) [reply azure-pipelines.yml:30]: This is filtered due to the issue you found in this PR.

1.0.0 codebase is using `.Csv("Resources\\people.csv");`

Which doesn't work on linux which requires `.Csv($"{TestEnvironment.ResourceDirectory}people.csv");` to be OS agnostic.

Comment (imback82) [reply azure-pipelines.yml:30]: Can you add a comment that this is due to the test issue, not the functionality changes? And you can add a link to this PR for the fix.

**Review by imback82 — APPROVED** [review]
LGTM, thanks @suhsteve!

**Comment by imback82** [comment]
@Niharikadutta could you add this to 1.0.1 release note?

**Pull request merged**